File: lesson-4/python/Program.py
import sys
from LightSm import LightSm

def read_input_run_sm(sm):
    print("\nCurrent state: " + LightSm.stateIdToString(sm.stateId))
    print("Please type 'i', 'd', 'o': ", end='')

    line = input().strip()
    if len(line) > 0:
        c = line[0]  # Get the first character
        run_sm_for_char(sm, c)
    else:
        run_sm_for_char(sm, '\0')


def run_sm_for_char(sm, c):
    valid_input = True
    event_id = LightSm.EventId.OFF

    if c == 'i':
        event_id = LightSm.EventId.INC
    elif c == 'd':
        event_id = LightSm.EventId.DIM
    elif c == 'o':
        event_id = LightSm.EventId.OFF
    else:
        valid_input = False

    if valid_input:
        print("Dispatching event: " + LightSm.eventIdToString(event_id))
        sm.dispatchEvent(event_id)
    else:
        print("Invalid input. Not running state machine.", file=sys.stderr)


if __name__ == "__main__":
    sm = LightSm()

    print("USAGE:")
    print("  type i <enter> to send INC event to state machine.")
    print("  type d <enter> to send DIM event to state machine.")
    print("  type o <enter> to send OFF event to state machine.")
    print("")
    print("Hit <enter> to start")
    input()  # Wait for user to press Enter

    sm.start()

    while True:
        read_input_run_sm(sm)

File: lesson-4/java/Program.java
import java.util.Scanner;
import my_light_package.LightSm;

public class Program {
    public static void main(String[] args) {
        LightSm sm = new LightSm();
        Scanner scanner = new Scanner(System.in);

        System.out.println("USAGE:");
        System.out.println("  type i <enter> to send INC event to state machine.");
        System.out.println("  type d <enter> to send DIM event to state machine.");
        System.out.println("  type o <enter> to send OFF event to state machine.");
        System.out.println("");
        System.out.println("Hit <enter> to start");
        scanner.nextLine(); // Waits for user to press Enter

        sm.start();

        while (true) {
            readInputRunSm(sm, scanner);
        }
    }

    private static void readInputRunSm(LightSm sm, Scanner scanner) {
        System.out.println("\nCurrent state: " + LightSm.stateIdToString(sm.stateId));
        System.out.print("Please type 'i', 'd', 'o': ");

        String line = scanner.nextLine();
        char c = '\0';

        if (line.length() > 0)
            c = line.charAt(0); // Get the first character

        runSmForChar(sm, c);
    }

    private static void runSmForChar(LightSm sm, char c) {
        boolean validInput = true;
        LightSm.EventId eventId = LightSm.EventId.OFF;

        switch (c) {
            case 'i':
                eventId = LightSm.EventId.INC;
                break;
            case 'd':
                eventId = LightSm.EventId.DIM;
                break;
            case 'o':
                eventId = LightSm.EventId.OFF;
                break;
            default:
                validInput = false;
                break;
        }

        if (validInput) {
            System.out.println("Dispatching event: " + LightSm.eventIdToString(eventId));
            sm.dispatchEvent(eventId);
        } else {
            System.err.println("Invalid input. Not running state machine.");
        }
    }
}
